### ISSUE TITLE
Azure Service Bus Queue Python receiver.reconnect() call causes queue message deletion

### ISSUE THREAD

**Issue by rswgnu** [Azure/azure-sdk-for-python#7079]
I have found in V0.50.1 using the Python SDK that when using ReceiveAndDelete mode on the receiver, merely calling the receiver's reconnect() method after a disconnection causes a dequeue and deletion of the most recently added queue item:

Queue len = 6
Reconnecting...
Queue len after reconnect = 5

Please fix this ASAP if possible.

**Comment by kaerm** [comment]
Thanks @rswgnu for reporting this our team will take a look at it!

**Comment by mayurid** [comment]
@annatisch could you take a.look

**Comment by annatisch** [comment]
Thanks @rswgnu  - I have written a short test to confirm this behaviour:
```
sb_client = ServiceBusClient.from_connection_string(conn_str, debug=True)
queue = sb_client.get_queue(standard_queue)

with queue.get_sender() as sender:
    for i in range(5):
        sender.send(Message(f"Message {i}"))

messages = queue.get_receiver(mode=ReceiveSettleMode.ReceiveAndDelete, idle_timeout=5)
batch = messages.fetch_next()
print(batch[0])
count = len(batch)

print("RECONENCTING")
messages.reconnect()
print("FINISHED RECONNECT")
for message in messages:
    print(message)
    count += 1
assert count == 5
```
And as you say - a message ("Message 1") always gets lost in the reconnect.
Output:
```
2019-09-18 07:28:09,791 uamqp.c_uamqp INFO     b'<- [ATTACH]* {receiver-link-c51cb0ab-f48e-4a7d-b4d3-1c07d994aefb,2,false,1,0,* {amqps://pysbtest.servicebus.windows.net/53b54470-b622-4d8c-8fbe-023b19669165,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL},* {SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9,NULL,NULL,NULL,NULL,NULL,NULL},NULL,NULL,0,1048576,NULL,NULL,NULL}'
2019-09-18 07:28:09,792 uamqp.c_uamqp INFO     b'-> [FLOW]* {2,65535,1,65535,2,0,1}'
2019-09-18 07:28:09,796 uamqp.receiver INFO     Message receiver b'receiver-link-c51cb0ab-f48e-4a7d-b4d3-1c07d994aefb' state changed from <MessageReceiverState.Opening: 1> to <MessageReceiverState.Open: 2> on connection: b'SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9'
2019-09-18 07:28:09,849 uamqp.c_uamqp INFO     b'<- [TRANSFER]* {2,1,<>,0,true,false,NULL,NULL,NULL,NULL,true}'
2019-09-18 07:28:09,850 uamqp.c_uamqp INFO     b'-> [FLOW]* {3,65534,1,65535,2,1,1}'
Message 0
RECONENCTING
2019-09-18 07:28:09,859 uamqp.c_uamqp INFO     b'-> [DETACH]* {2,true}'
2019-09-18 07:28:09,860 uamqp.connection INFO     Shutting down connection b'SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9'.
2019-09-18 07:28:09,863 uamqp.c_uamqp INFO     b'-> [DETACH]* {0,true}'
2019-09-18 07:28:09,864 uamqp.c_uamqp INFO     b'-> [DETACH]* {1,true}'
2019-09-18 07:28:09,875 uamqp.c_uamqp INFO     b'-> [END]* {}'
2019-09-18 07:28:09,877 uamqp.c_uamqp INFO     b'-> [CLOSE]* {}'
2019-09-18 07:28:09,879 uamqp.connection INFO     Connection shutdown complete b'SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9'.
2019-09-18 07:28:10,131 uamqp.c_uamqp INFO     b'-> Header (AMQP 0.1.0.0)'
2019-09-18 07:28:10,185 uamqp.c_uamqp INFO     b'<- Header (AMQP 0.1.0.0)'
2019-09-18 07:28:10,186 uamqp.c_uamqp INFO     b'-> [OPEN]* {SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9,pysbtest.servicebus.windows.net,65536,65535,NULL,NULL,NULL,NULL,NULL,{[product:servicebus.python],[version:0.50.1],[framework:Python 3.7.4],[platform:win32]}}'
2019-09-18 07:28:10,240 uamqp.c_uamqp INFO     b'<- [OPEN]* {03c263b2d4f1468e9d45485821064836_G31,NULL,65536,4999,240000,NULL,NULL,NULL,NULL,NULL}'
2019-09-18 07:28:10,241 uamqp.c_uamqp INFO     b'-> [BEGIN]* {NULL,0,65536,65536,4294967295}'
2019-09-18 07:28:10,298 uamqp.c_uamqp INFO     b'<- [BEGIN]* {0,1,5000,65536,255,NULL,NULL,NULL}'
...
2019-09-18 07:28:10,480 uamqp.c_uamqp INFO     b'-> [ATTACH]* {receiver-link-dade5557-9f18-4d23-9b95-9619e10cbd3a,2,true,1,0,* {amqps://pysbtest.servicebus.windows.net/53b54470-b622-4d8c-8fbe-023b19669165},* {SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9},NULL,NULL,NULL,1048576}'
2019-09-18 07:28:10,592 uamqp.c_uamqp INFO     b'<- [ATTACH]* {receiver-link-dade5557-9f18-4d23-9b95-9619e10cbd3a,2,false,1,0,* {amqps://pysbtest.servicebus.windows.net/53b54470-b622-4d8c-8fbe-023b19669165,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL},* {SBReceiver-eb27c68a-2d84-4b8b-b888-8bb131b881e9,NULL,NULL,NULL,NULL,NULL,NULL},NULL,NULL,0,1048576,NULL,NULL,NULL}'
2019-09-18 07:28:10,594 uamqp.c_uamqp INFO     b'-> [FLOW]* {2,65535,1,65535,2,0,1}'
FINISHED RECONNECT
2019-09-18 07:28:10,654 uamqp.c_uamqp INFO     b'<- [TRANSFER]* {2,1,<>,0,true,false,NULL,NULL,NULL,NULL,true}'
2019-09-18 07:28:10,654 uamqp.c_uamqp INFO     b'-> [FLOW]* {3,65534,1,65535,2,1,1}'
Message 2
2019-09-18 07:28:10,712 uamqp.c_uamqp INFO     b'<- [TRANSFER]* {2,2,<>,0,true,false,NULL,NULL,NULL,NULL,true}'
2019-09-18 07:28:10,713 uamqp.c_uamqp INFO     b'-> [FLOW]* {4,65533,1,65535,2,2,1}'
Message 3
2019-09-18 07:28:10,766 uamqp.c_uamqp INFO     b'<- [TRANSFER]* {2,3,<>,0,true,false,NULL,NULL,NULL,NULL,true}'
2019-09-18 07:28:10,767 uamqp.c_uamqp INFO     b'-> [FLOW]* {5,65532,1,65535,2,3,1}'
Message 4
2019-09-18 07:28:15,853 uamqp.client INFO     Timeout reached, closing receiver.
```
However the client is showing no indication that a frame was ever received during the reconnect process. This could indicate that the service is sending a frame that the client either never receives or it's being immediately discarded because the connection is already in a shutdown state.
I will investigate whether we can improve the behaviour here, however please bear in mind that loss of messages during an connection interruption will _always_ be a risk when using ReceiveAndDelete mode. So if receipt of every message is important, please use PeekLock mode.

**Comment by rswgnu** [comment]
> it's being immediately discarded because the connection is already in a shutdown state.

Could you explain the above further?

I understand if something like a power outage happens when a client is trying to read a message from a queue that the message may be lost but the queue should not be dropping messages prior to read attempts being made; it should be persistent subject to whatever "time-to-live" policy is given for the queue (regardless of the state of a client).  ServiceBus disconnects clients pretty rapidly in my experience so reconnects are needed all the time and should not affect any items in the queue.  Do you agree with this?

There was a specific reason we chose not to use PeekLock mode though I do not recall it right now.

**Comment by annatisch** [comment]
Thanks @rswgnu - yes I agree - I would think dropping messages due to disconnect/reconnect should be an exceptional case rather than normal behaviour - so I certainly need to do a deeper investigation into the C library to ensure that incoming messages are not being needlessly discarded (as opposed to discarded because they were received erroneously).

**Comment by rswgnu** [comment]
Ok, great.  Also, I am pretty sure I verified that the receiver is not in a shutdown state yet the message is still dropped whenever a reconnect is done.  If there were a way we could keep connections open for hours at a time as we would like to do (because we would only have a few clients accessing our queue at a time), then that would help us as well and we wouldn't need the reconnects generally.  Is there any way to do that?

**Comment by annatisch** [comment]
A reconnect is generally performed due to an error in one of the connection/link/authentication - so keeping the connection open is not always an option. Therefore if a reconnect is performed, the receiver would have been in a shutdown state at some point... though it possible that state has not been surfaced to the client due to some kind of error.
Do you know under what circumstances a reconnect is being performed in your case?

**Comment by rswgnu** [comment]
I am manually calling reconnect after the receiver is forcibly disconnected from the queue.  Prior to that I was just recreating the receiver to dequeue each item and that did not work properly either as I recall.  The disconnection is incredibly fast relative to the rate at which we want to dequeue items.  Our processes run for hours and we don't want to dequeue everything upfront because then we would just have to create a local persistent store.

**Comment by annatisch** [comment]
Ah that's interesting.
I confess that the `reconnect` function was never intended to be manually called - however that's on me as I never explicitly marked it as internal either...
Do you notice that messages are still dropped if you use your previous implementation of recreating the receiver each time? I would suspect that the behaviour should be the same and am very curious as to what you observed.

**Comment by rswgnu** [comment]
Yes, that is why I then tried using reconnect.  Personally, I think it would be handy to have as part of the API because why do we want to recreate the receiver locally just because the other end of the connection dropped?  Reconnect should be lighter and faster.

**Comment by rswgnu** [comment]
Hi Anna:

What is the status.  We need some resolution on this ASAP as it is a blocking bug.  Please let us know if someone is working on this on a daily basis or not.  Thanks.  Full sample code with a valid workaround would work as well.

**Comment by annatisch** [comment]
Hi @rswgnu - yes I am currently working on this alongside a couple of other open issues for the Service Bus library.
Though I would recommend using PeekLock mode as a short term work around for this issue - you mentioned that there's a specific reason you can't, so it would be good to hear more about this to determine whether there's something else that needs to be resolved.

**Comment by rswgnu** [comment]
An I just reopen the receiver connection each time I want to dequeue a message and avoid using reconnect?  If not, I will have to revert to PeekLock and all that entails.

-- Bob

**Comment by annatisch** [comment]
I don't think so - I believe the missing message is happening during either shutting down, or opening a new connection. So either way this is achieved (reconnect or reopen) I think the same behaviour would occur. Therefore I would recommend PeekLock until I can isolate where the missing bytes are going...

**Comment by annatisch** [comment]
Hi @rswgnu -I have finally isolated the root cause of this issue and am working on a fix for it now :)

**Comment by annatisch** [comment]
I have a PR open with a potential fix here:
https://github.com/Azure/azure-uamqp-python/pull/96

And preview builds of uAMQP can be download here:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=119291

The fix for this issue requires no changes to the Service Bus SDK.
So it would be great if you could try out the updated uAMQP to see whether it resolves the issue for you.

**Comment by rswgnu** [comment]
Could you provide a step-by-step description for substituting your uAMQP version for the standard one in the SDK, so I don't have to figure it out and can test this faster?  Thanks.

**Comment by annatisch** [comment]
Sure, so start by downloading the uAMQP build according to your platform from below:
Latest build: https://dev.azure.com/azure-sdk/public/_build/results?buildId=121096

Then you should simply be able to install it into your test/dev/local environment:
```
pip install --upgrade path/to/local/uamqp.whl
```
Then you should simply be able to re-run your existing tests/code that would have previously reproduced the bug.

**Comment by rswgnu** [comment]
With the new uamqp library, as long on the Azure Service Bus queue receiver connection stays open, then dequeing now works properly.  But of course the connection closes fairly rapidly so we need to know the lowest-cost code to re-establish the connection given a prior good handle for the receiver.  Calling get_receiver again (creating a new receiver object) seems to work but we really want something like the reconnect you exposed previously.  How should we handle this?

**Comment by rswgnu** [comment]
Could you get this fix into production quickly so we could just update through normal pip channels?  We deploy to multiple virtual machines and share our Python virtual environments, so it would be a bit of a mess to install a custom library into our production servers.

**Comment by annatisch** [comment]
Great - thanks for the feedback @rswgnu!
The `reconnect` function you were using previously should now also be fixed with this patch.
I will not be "unexposing" the reconnect functionality, and have now added test cases for this scenario - so you should be able to use `reconnect` again.

All going well we plan to have this update published to PyPI early next week.

**Comment by rswgnu** [comment]
Sounds good.  I will test reconnect tomorrow.

-- Bob

**Comment by rswgnu** [comment]
Hi Anna:  reconnect() is not working if the only change I install is your updated uamqp code.  I get a connection is in unknown state after doing a reconnect().  Please explain any other changes that need to be applied and show a simple use of reconnect() where you wait for 5 minutes (sleep(300)) and then do a reconnect().

**Comment by rswgnu** [comment]
Here is my debug output:
INFO:__main__:Task # 0 "old" command, uuid: 63e169d2-e52b-11e9-bdde-f43909d9b385: enqueued
Enqueued Task # 0 "old" command: Q Size: 1; Task Addr: 140668735774560; JSON: {"cmd_type": "old", "cmds": ["a", "b", "c"], "output_file": null, "number": 0, "at": null, "at_start": null, "at_end": null, "uuid": "63e169d2-e52b-11e9-bdde-f43909d9b385"}
INFO:__main__:Task # 1 "old" command, uuid: 644891ac-e52b-11e9-bdde-f43909d9b385: enqueued
Enqueued Task # 1 "old" command: Q Size: 2; Task Addr: 140668735773440; JSON: {"cmd_type": "old", "cmds": ["b", "c", "d"], "output_file": null, "number": 1, "at": null, "at_start": null, "at_end": null, "uuid": "644891ac-e52b-11e9-bdde-f43909d9b385"}
INFO:__main__:Task # 2 "old" command, uuid: 64aff23e-e52b-11e9-bdde-f43909d9b385: enqueued
Enqueued Task # 2 "old" command: Q Size: 3; Task Addr: 140668735774112; JSON: {"cmd_type": "old", "cmds": ["c", "d", "e"], "output_file": null, "number": 2, "at": null, "at_start": null, "at_end": null, "uuid": "64aff23e-e52b-11e9-bdde-f43909d9b385"}
INFO:__main__:Task # 3 "old" command, uuid: 65173818-e52b-11e9-bdde-f43909d9b385: enqueued
Enqueued Task # 3 "old" command: Q Size: 4; Task Addr: 140668735773216; JSON: {"cmd_type": "old", "cmds": ["d", "e", "f"], "output_file": null, "number": 3, "at": null, "at_start": null, "at_end": null, "uuid": "65173818-e52b-11e9-bdde-f43909d9b385"}
INFO:__main__:Task # 4 "old" command, uuid: 657da31e-e52b-11e9-bdde-f43909d9b385: enqueued
Enqueued Task # 4 "old" command: Q Size: 5; Task Addr: 140668735773776; JSON: {"cmd_type": "old", "cmds": ["e", "f", "g"], "output_file": null, "number": 4, "at": null, "at_start": null, "at_end": null, "uuid": "657da31e-e52b-11e9-bdde-f43909d9b385"}
INFO:__main__:Task # 5 "old" command, uuid: 65ee15ea-e52b-11e9-bdde-f43909d9b385: enqueued
Enqueued Task # 5 "old" command: Q Size: 6; Task Addr: 140668735157360; JSON: {"cmd_type": "old", "cmds": ["f", "g", "h"], "output_file": null, "number": 5, "at": null, "at_start": null, "at_end": null, "uuid": "65ee15ea-e52b-11e9-bdde-f43909d9b385"}
Queue len = 6
INFO:__main__:  Task # 0 "old" command, uuid: 63e169d2-e52b-11e9-bdde-f43909d9b385: dequeued
Dequeued Task # 0 "old" command: Q Size: 5; Task Addr: 140668735772432; JSON: {"cmd_type": "old", "cmds": ["a", "b", "c"], "output_file": null, "number": 0, "at": null, "at_start": null, "at_end": null, "uuid": "63e169d2-e52b-11e9-bdde-f43909d9b385"}
Pausing for 5 minutes to check handling after timeouts...
Queue len = 5
WARNING:uamqp.connection:ConnectionClose('ErrorCodes.UnknownError: Connection in an unexpected error state.')
Trying to dequeue when queue is empty...
Queue len = 5
Reconnecting...
Queue len after reconnect = 4
Queue len after _used cleared = 4

**Comment by annatisch** [comment]
Thanks @rswgnu - can I confirm whether the code you're running looks roughly similar to the test I have here:
https://github.com/Azure/azure-sdk-for-python/pull/7453/files#diff-7c1b8fc70db3a3f3f724231ff9a11da5R62

**Comment by rswgnu** [comment]
Yes, fairly similar with a longer delay between reconnects.

**Comment by rswgnu** [comment]
Can you replicate the reconnect problem?  Any movement on it yet?

**Comment by annatisch** [comment]
No - I am unable to reproduce any further issues here.
I bumped the delay to 5 minutes but the test still passed.
Current plan is to release the fix as it is early next week and we will reassess status at that time.

**Comment by rswgnu** [comment]
Ok, thanks.

**Comment by annatisch** [comment]
The updated uAMQP (v1.2.3) is now released. You should be able to install with:
```
$ pip install --upgrade uamqp
```

**Comment by rswgnu** [comment]
That’s very good news.

-- Bob

**Comment by mayurid** [comment]
@rswgnu : Looking forward to your feedback on this

**Comment by rswgnu** [comment]
Creating a new receiver object with each get/read works properly now but if I reuse the receiver object, it times out and I can’t find a working method to reconnect.  The reconnect() method does not work after the time out.

Please send the most basic reconnect code you can, using a time.sleep(300) delay between reconnects and dequeues and I will test that.

Thanks,

-- Bob

**Comment by rswgnu** [comment]
See my prior comment.  Can we get some followup on this so we don't have to create a new receiver object every time we dequeue a message?  Thanks.

**Comment by engebret** [comment]
Has there been any resolution on this one?  I believe we are seeing something similar.  I've moved to uamqp v1.2.5 but still seeing exceptions after a connection has gone unused for a few minutes.

**Comment by KieranBrantnerMagee** [comment]
Hey folks,
First let me apologize that this sat for so long.
Second, to address the (what I believe are) two distinct points here.
1. receiver reuse.
  Unfortunately in the 0.50.* versions, reuse of the push handler (`for message in receiver` vs the pull handler `fetch_next`) was explicitly not allowed due to issues with connection/credential state.  This has been mitigated in the version of this lib pending GA, currently in preview (7.0.0b2 latest), you will be able to call `for message in receiver` again even if it was already called and exited prior.  That said, I'd still be very surprised that creating a new receiver for _each_ message would be required.  In the occasion that the receiver was terminated for whatever reason, would it not be sufficient to fetch another in the error handler?  E.g. a pattern as such:
```python
while True:
    with queue_client.get_receiver() as receiver:
        try:
            for message in receiver:
                time.sleep(5) #some arbitrary processing
                message.complete()
        except:
            continue
```
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A similar approach should be workable using fetch_next for the inner loop, if that pattern is 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;desired.  Don't hesitate to let me know if I've missed the motivation for why creating a new 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;receiver per-message would be needed as a workaround, since as I understand it anna's fix 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;addressed the dropped-message issue.

2. unused connection expiry.
&nbsp;&nbsp;I'm _hoping_ that this is distinct, and merely incurring the pain of item 1 when it occurs.  
&nbsp;&nbsp;(Rooted in a service-side disconnect at 10 minutes which we mitigated in a recent patch release) 
&nbsp;&nbsp;We've been triaging this with engebret across a few other issues, see #9144 and [uamqp #135](https://github.com/Azure/azure-uamqp-python/issues/135)

**Comment by KieranBrantnerMagee** [comment]
Just wanted to mention that we've officially released ServiceBus 0.50.3 which contains the aformentioned fixes for the long-term service side connection disconnect.

I would hope between this and the aformentioned call patterns the issues here are mitigated; if there's any recurrence of this or further questions don't hesitate to reopen and/or reach out.
